onnxruntime/core/optimizer/slice_elimination.cc

```diff
@@ -29,23 +29,94 @@ bool EliminateSlice::SatisfyCondition(const Graph& graph, const Node& node) cons
+    };
+
+    // Returns a pointer to the initializer if it is constant; otherwise, a nullptr.
+    auto get_initializer_if_constant =
+        [&graph, get_input_if_exists](size_t input_idx) -> const ONNX_NAMESPACE::TensorProto* {
+      const NodeArg* input = get_input_if_exists(input_idx);
+      return input ? graph_utils::GetConstantInitializer(graph, input->Name()) : nullptr;
+    };
+
+    auto get_initializer_data =
+        [](const ONNX_NAMESPACE::TensorProto* initializer) -> std::vector<int64_t> {
+      Initializer init(*initializer);
+      if (initializer->data_type() == ONNX_NAMESPACE::TensorProto::INT32) {
+        int32_t* init_data = init.data<int32_t>();
+        return std::vector<int64_t>(init_data, init_data + init.size());
+      } else if (initializer->data_type() == ONNX_NAMESPACE::TensorProto::INT64) {
+        int64_t* init_data = init.data<int64_t>();
+        return std::vector<int64_t>(init_data, init_data + init.size());
+      }
+      return {};
+    };
+
+    // Starts and ends inputs have to exist, be constant, and be of the same size.
+    const ONNX_NAMESPACE::TensorProto* starts_init = get_initializer_if_constant(1);
+    const ONNX_NAMESPACE::TensorProto* ends_init = get_initializer_if_constant(2);
+    if (starts_init && ends_init) {
+      starts = get_initializer_data(starts_init);
+      ends = get_initializer_data(ends_init);
+
+      if (starts.size() == 0 || ends.size() == 0 || starts.size() != ends.size()) {
+        return false;
+      }
+
+      // If axes input exists, it should be constant and of the same size as starts/ends.
+      if (get_input_if_exists(3)) {
+        const ONNX_NAMESPACE::TensorProto* axes_init = get_initializer_if_constant(3);
+        if (!axes_init || axes_init->dims_size() != 1 ||
+            static_cast<size_t>(axes_init->dims().Get(0)) != starts.size()) {
+          return false;
+        }
+
+        // If steps input exists, it should be constant and all values should be 1.
+        if (get_input_if_exists(4)) {
+          const ONNX_NAMESPACE::TensorProto* steps_init = get_initializer_if_constant(4);
+          if (!steps_init) {
+            return false;
+          }
+          std::vector<int64_t> steps = get_initializer_data(steps_init);
+          if (steps.size() != starts.size()) {
+            return false;
+          }
+          for (int64_t step : steps) {
+            if (step != 1) {
+              return false;
+            }
+          }
+        }
+      }
+    } else {
+      // Should be unreachable, but just to be safe in case a new op version is added.
+      return false;
     }
-  } else if (axes.size() != starts.size()) {
-    return false;
   }
 
   // For now eliminate slice operators if starts=0 and ends=MAX_INT.
   // TODO: Take into account the input's shape to get a tighter bound for the ends.
-  for (size_t i = 0; i < axes.size(); ++i) {
+  for (size_t i = 0; i < starts.size(); ++i) {
```

Comment:
Is there an existing bug in that nowhere seems to check the axes values? We can't eliminate the Slice if the axes value is provided but has out of order values can we?

Reply:
Semantics of slice is not clear to me...
Reading the documentation, I don't understand if Slice is allowed to reorder the axes of the input array. It is allowed to have the axes be out of order, like [1,0] (btw no duplicates are allowed in our shape inference), but that does not mean reordering should happen. Documentation simply says that it is the same as the numpy slice, and in that there is no syntax that allows reordering.

Reply:
I think you're right - it's not for re-ordering so we're good